packages/@aws-cdk/cx-api/lib/private/flag-modeling.ts

```diff
@@ -0,0 +1,39 @@
+export enum FlagType {
+  /**
+   * Change the default behavior of the API
+   *
+   * The old behavior is not disrecommended, and possible to achieve with source
+   * code changes. Also valid for changes that don't affect CloudFormation, but
+   * the CXAPI contract.
+   */
+  ApiDefault,
+
+  /**
+   * Address a bug/introduce a recommended change
+   *
+   * The old behavior is no longer recommended. The only way to achieve it is by
+   * keeping the flag at the legacy value.
+   */
+  BugFix,
+
+  /**
+   * Advertise the presence of this context option in `cdk.json`
+   */
+  VisibleContext,
+};
+
+export interface FlagInfo {
+  /** Flag type */
+  readonly type: FlagType;
+  /** Single-line description for the flag */
+  readonly summary: string;
+  /** Detailed description for the flag */
+  readonly details: string;
+  /** Version number the flag was introduced in each version line */
+  readonly introducedIn: { v1?: string; v2?: string };
+  /** Default value in the given major version */
+  readonly defaults?: { v2?: any };
+  /** Default in new projects */
+  readonly recommendedValue: any;
```

Comment:
Which is exactly what it is! (Well, roughly) In fact, I had exactly what you suggested at first, but I ultimately thought this was easier to understand. Consider, for a hypothetical flag, the default value for `v2` is `true`:

* Should I, the human editing this info, now set `defaultValueForNewProjects: true`, or leave it out? How does that relate to what `cdk init` is going to do?
* With the current naming, the `recommendedValue = true`. Both of these are simple statements of fact. `cdk init` can later on interpret that `default == recommended`, and that nothing needs to be done.

Reply:
Yeah I guess I can see that, but I think there needs to be more documentation explaining this. As a user adding a new flag, I'm not sure what I should set where. For example, what if we introduce a new flag like the `newStyleStackSynthesis` flag. We would set `recommendedValue: false` & `defaults: { v2: true }`. It reads like we are not _recommending_ this flag when we actually are recommending it, we just aren't setting it as the default for new projects.

Reply:
> newStyleStackSynthesis is recommended: true, not recommended: false.
If in fact it was true that we had a flag with recommended: false, default: true, then it would be added to cdk.json with the value false (seems correct?).

I think my confusion was because I was thinking from the point of view of having a `v3` (i.e. recommendedValue might be false for `v2` and `true` for `v3`), but that's getting ahead of myself 😂.